Serializer

joanromano opened the pull request.

MP0w committed:
ADD serializer

MP0w committed:
MOD edited access scopes

MP0w committed:
ADD unit tests
FIX PropertyPolicy serialization

joanromano commented on Source/Serializer.swift:
```diff
@@ -0,0 +1,77 @@
+    if let value = value as? KakapoSerializable {
+        return value.serializeChildren()
+    } else {
+        assert(value is AnyObject) // TODO: throw
```
Make this function throw?

joanromano commented on Source/PropertyPolicy.swift:
```diff
@@ -0,0 +1,28 @@
+        if let obj = obj {
+            return Mirror(reflecting: obj)
+        }
+        return Mirror(obj, children: [])
```
As previously asked, why we need different Mirror inits? Probably good to add comments here for that

MP0w replied:
Would love to do a default implementation that doesnt require this... But associatedType... This just means nil... Mirror with no childrens

joanromano replied:
Sorry don't understand what you say :S Where is the associated type? What does "This just means nil" mean?

joanromano commented on KakapoExample/KakapoTests/SerializerTests.swift:
```diff
@@ -0,0 +1,126 @@
+                let empty = MaybeEmpty(maybeIgnored: IgnorableNilProperty<Int>(obj: 1))
+                let serialized = serialize(empty) as? [String: Int]
+                expect(serialized?["maybeIgnored"]).to(be(1))
+            }
```
Two last describe fail and the optional serialization cast crashes: probably should re map those dictionaries?

zzarcon commented:
@MP0w ping

MP0w committed:
Refactored property policies
Renamed some types
Test for recursive Optionals and Policies serialization are failing

MP0w committed:
Swift 2.2 (finally using Xcode 7.3)

MP0w committed:
FIX recursive Optional Serialization
ADD recursive PropertyPolicy initialization

MP0w committed:
MOD remove dead code

MP0w committed:
Merge branch 'master' into feature/serializer

# Conflicts:
#	KakapoExample/KakapoExample.xcodeproj/project.pbxproj
#	KakapoExample/KakapoExample.xcodeproj/xcshareddata/xcschemes/Kakapo.xcscheme
#	KakapoExample/KakapoExample.xcodeproj/xcshareddata/xcschemes/KakapoExample.xcscheme
#	KakapoExample/Pods/Pods.xcodeproj/project.pbxproj

MP0w committed:
FIX travis schemes

MP0w committed:
REM unit test target from travis

MP0w committed:
Downgrade Xcode version

MP0w committed:
swift <2.2 compatibility

joanromano commented on Source/Serializer.swift:
```diff
@@ -0,0 +1,83 @@
+    if let value = value as? Serializable {
+        return value.serialize()
+    } else {
+        assert(value is AnyObject) // TODO: throw
```
Why do we want to assert/throw? For String objects this is actually where we want to reach: returning the String, right?

MP0w replied:
String would work because is bridgeable. We have this constrain because only a dictionary of Anyobject can be converted to json

joanromano replied:
Actually for Strings, Int, Float, Double, etc this is where we want to reach right?

joanromano replied:
Then, could you please add a test that tests this assertion? I guess a struct that does not conform to Serializable, since other value types are automatically bridged?

joanromano replied:
Or make it `throw` as DB method `func insert<T: KStorable>(object: T) throws` and test that

joanromano commented on Source/PropertyPolicy.swift:
```diff
@@ -0,0 +1,40 @@
+            return object
+        }
+    }
+}
```
Just out of curiosity: how would a property policy that wants to, for instance, serialize nil as 0 work? The object proerty will have to return 0 or 1 based on his value?

I think the documentation on how to create custom PropertPolicty is quite crucial, otherwise is a bit  confusing.

MP0w replied:
You would just return 0 if nil and the object if not nil, type must be Any if the object is not Int

joanromano replied:
Meaning that your representation of `object` has to be something like this?

```
public var object: Any {
   return _someInternalProperty == nil ? 0 : _someInternalProperty
}
```

MP0w replied:
Yes exactly

joanromano commented on KakapoExample/KakapoTests/SerializerTests.swift:
```diff
@@ -0,0 +1,155 @@
+            }
+
+            it("recursively serialize IgnorableNilProperties") {
+                let notEmpty = MaybeEmpty(IgnorableNilProperty(IgnorableNilProperty(1)))
```
Wow

MP0w replied:
Doesn't make sense for IgnorableNil but might make sense for propertyPolicy with different logic

joanromano commented on KakapoExample/KakapoTests/SerializerTests.swift:
```diff
@@ -0,0 +1,155 @@
+            }
+            
+            it("recursively serialize Optionals") {
+                let optional = MaybeEmpty(Optional.Some(Optional.Some(1)))
```
Wow(x2)

joanromano commented on KakapoExample/KakapoTests/SerializerTests.swift:
```diff
@@ -0,0 +1,155 @@
+                expect(serialized?["value"]).to(be(1))
+            }
+        }
+    }
```
The tests for optionals and IgnorableNilProperty are really good, though don't you think we need deeper tests for arrays and dictionaries? I am missing some tests with numerous items, both for arrays and dictionaries.

MP0w replied:
Oh, do you think we should at least have 2?

joanromano replied:
I mean arrays with more items, same with dictionaries

joanromano commented:
:+1: fine for me to merge after the proposed changes

MP0w committed:
ADD additional tests for Array and Dictionaries
